### ISSUE TITLE
Handle enabling the feature flag in the Eventing component reconciler

### ISSUE THREAD

**Issue by pxsalehi** [kyma-project/kyma#13321]
Once we merge the JetStream feature flag in #13251, we also need to address how this can be enabled using the Kyma reconciler. As of https://github.com/kyma-project/kyma/pull/13259, I have tested the Kyma reconciler using Kyma CLI unstable version `main-9feb3a3`.

- Upgrade from `main` (https://github.com/kyma-project/kyma/commit/75da62ce99a9bc773759afb3e76da28de715f574, a 2.x version w/o the feature flag) to `PR-13259` (with feature flag enableJetStream=false): works w/o any reconciler changes required. The only place the feature flag shows up is in the eventing-controller deployment env vars.

- Installing `main` (enableJetStream=true): works w/o any reconciler changes required. It also correctly brings up a JetStream cluster.

- Upgrading from  `main` (enableJetStream=false) to `main` (enableJetStream=true): It correctly updates the eventing-controller deployment and the NATS ConfigMap (`eventing-nats-config`), however, you'd need to restart the NATS servers (or the StatefulSet scaled to 0 and back up) to have JetStream enabled. I think, upgrading from `main` w/o any feature flag directly to enabled feature flag would also require the restart, as the NATS cluster is already running.

Depending on how and when we plan to release the JetStream backend, most likely we would require some pre-action in the eventing component reconciler that would perform the NATS cluster restart.

**Comment by VladislavPaskar** [comment]
Unfortunately I couldn't reproduce  the `false->true` upgrade scenario, but found out that the JS-specific data is not being removed from the Nats StatefulSet and Pod(s) during the `true->false` scenario. I added a PR, which should fix that problem: https://github.com/kyma-incubator/reconciler/pull/864

**Comment by pxsalehi** [comment]
Which part couldn't you reproduce? You mean after enabling the flag, no restart of the NATS server (i.e. recreating the pod) was required? If that is the case, then we would not need any change on the reconciler.

This issue is about enabling the flag. Do we need to support disabling it?

Why are you removing the StatefulSet? I think removing the pod, or scaling down and then up should be enough.

**Comment by VladislavPaskar** [comment]
I couldn't reproduce the part when provisioning from `enableJetStream=false` to `enableJetStream=true`, which is the third bullet point in the issue. In that case provisioner was adding the Jetstream related data to the StatefulSet (in our case that is the js secret to the pod's envs). 

I understand that this issue is about enabling, but I thought we could add it as a part of this issue, since it's all about toggling the JS feature.

I remove the statefulSet and and not scaling it down, because it's the statefulSet which has the wrong data (after disabling the JS flag) and not the pods, so scaling down wouldn't help, it needs to be removed.

**Comment by pxsalehi** [comment]
> I couldn't reproduce the part when provisioning from enableJetStream=false to enableJetStream=true ...
Does this mean that the NATS server just switched to JetStream enabled without any restart? How did you verify that the NATS server is now running with JetStream enabled?

> I understand that this issue is about enabling, but I thought we could add it as a part of this issue, since it's all about toggling the JS feature.
No, it is not all about toggling. It is only about enabling. You can discuss disabling with the team, and see if it is needed.

>  remove the statefulSet and and not scaling it down, because it's the statefulSet which has the wrong data...
I'm not sure, if I entirely follow, but the pod is created based on the template in the statefulset. When the reconciler updates/patches the statefulset, creating a new pod will follow the new pod template in the new statefulset. You don't need to remove the statefulset, to "update" the pod template, I think.

For now, let's put this on hold. I will look into another approach where we would probably not need any reconciler changes.

**Comment by VladislavPaskar** [comment]
Ok, let's block it until you are done if tour approach and I will create another issue for disabling and will will see in the grooming if you need it in the first place.

**Comment by k15r** [comment]
Lets handle the investigation and potential implementation of the config reloader also in this ticket. (Unblocking)

**Comment by nachtmaar** [comment]
ToDos left:
- [ ] test again upgrade case 3
- [ ] config reloader

**Comment by nachtmaar** [comment]
I tested **upgrade case 3**: (Upgrading from main (enableJetStream=false) to main (enableJetStream=true)

**TL;DR**
- `fails` then using `Kyma CLI`
- `works` then using `kyma reconciler only`


# Case Upgrading with Kyma CLI
- initial: `kyma-stable deploy -s main --value global.features.enableJetStream=false`
- upgrade: `kyma-stable deploy -s main --value global.features.enableJetStream=true`

- Upgrade fails, eventing-controller is in crashbackloop

```json
{"level":"ERROR","timestamp":"2022-03-10T17:44:42Z","caller":"jetstream/reconciler.go:51","message":"start reconciler failed","context":{"name":"jetstream-subscription-reconciler","error":"nats: no responders available for request"}}
panic: nats: no responders available for request
```


Restarting the nats-pod fixes the problem
`k delete pod -n kyma-system eventing-nats-0`

# Case upgrading with mothership-reconciler
Restart **works** as expected, no manual restart required. I tested it twice :)

- event-nats-o pod was restarted
  - before upgrade: 2022-03-10T14:08:33Z
  - after upgrade: 2022-03-10T16:29:13Z
- jetstream was enabled
- upgrade took ˜ 5 minutes (nats pod not ready for some time and  eventing-controller crashing because of no connection to nats)


# How I tested:

## Installing kyma main

```shell
$ kyma-provision-gardener-azure-prod ns213321

$ kyma-stable version
Kyma CLI version: stable-9407a4a
Kyma cluster version: main

$ kyma-stable deploy -s main --value global.features.enableJetStream=false
- No previous Kyma version found
- Using Kyma from the workspace directory: /Users/xxx/.kyma/sources/main
- Initial setup
- Component 'CRDs' deployed
- Component 'cleaner' deployed
- Component 'certificates' deployed
- Component 'istio-configuration' deployed
- Component 'cluster-essentials' deployed
- Component 'kiali' deployed
- Component 'serverless' deployed
- Component 'monitoring' deployed
- Component 'rafter' deployed
- Component 'tracing' deployed
- Component 'helm-broker' deployed
- Component 'application-connector' deployed
- Component 'ory' deployed
- Component 'cluster-users' deployed
- Component 'service-catalog-addons' deployed
- Component 'eventing' deployed
- Component 'logging' deployed
- Component 'istio-resources' deployed
- Component 'api-gateway' deployed
- Component 'service-catalog' deployed
- Kyma deployed successfully!

Kyma is installed in version:   main
Kyma installation took:         0 hours 7 minutes

```

## Test eventing works

```shell
$ kyma-appnpair-app ns213321.kymatunas.shoot.canary.k8s-hana.ondemand.com
$ kyma-appnpair-events ns213321.kymatunas.shoot.canary.k8s-hana.ondemand.com
$ kyma-appnpair-events-bind ns213321.kymatunas.shoot.canary.k8s-hana.ondemand.com
$ kyma-appnpair-function ns213321.kymatunas.shoot.canary.k8s-hana.ondemand.com
$ kyma-appnpair-subscription ns213321.kymatunas.shoot.canary.k8s-hana.ondemand.com
$ kyma-send-ce-structured ns213321.kymatunas.shoot.canary.k8s-hana.ondemand.com
< HTTP/2 204
```

## Validate jetstream is not enabled

```shell
# check that jetstream section is not inside the config (yet)
$ k get cm -n kyma-system eventing-nats-config -oyaml | grep -C 5 jetstream
empty

# ENABLE_JETSTREAM_BACKEND env var is not set on the eventing-controller (yet)
$ k get deployments.apps -n kyma-system eventing-controller -oyaml |grep -i -A 1 ENABLE_JETSTREAM_BACKEND
        - name: ENABLE_JETSTREAM_BACKEND
          value: "false"

# Check sts events (we need this to make sure a new pod is created when enabling jetstream) 
$ k describe sts -n kyma-system  eventing-nats|grep -A 10 Events
Events:
  Type    Reason            Age   From                    Message
  ----    ------            ----  ----                    -------
  Normal  SuccessfulCreate  30m   statefulset-controller  create Pod eventing-nats-0 in StatefulSet eventing-nats successful


# check start time of nats pod (to make sure it got recreated later)
$ k get pods -n kyma-system eventing-nats-0 -o jsonpath='{.status.startTime}'
2022-03-10T14:08:33Z%

# cannot use nats CLI to show stream info (yet)
$ kubectl -n kyma-system port-forward svc/eventing-nats 4222
$ nats stream ls
nats: error: could not list streams: nats: no responders available for request, try --help
```

## Upgrade eventing to use jetstream

```shell
$ kyma-rec-mothership local --components eventing --value global.features.enableJetStream=true
...
{"level":"WARN","time":"2022-03-10T17:26:02.769+0100","message":"no configuration file found: set environment variable $RECONCILER_CONFIG or define it as CLI parameter"}
{"level":"INFO","time":"2022-03-10T17:26:02.769+0100","message":"Local installation started with kubeconfig /Users/xxx/SAPDevelop/kubeconfigs/kyma-13321"}
{"level":"INFO","time":"2022-03-10T17:26:02.769+0100","message":"Cloning GIT repository 'https://github.com/kyma-project/kyma' with revision 'main' into workspace '.workspace/main'"}
{"level":"INFO","time":"2022-03-10T17:26:02.769+0100","message":"Not able to create an In Cluster Client"}
{"level":"INFO","time":"2022-03-10T17:26:28.104+0100","message":"Starting local scheduler"}
{"level":"INFO","time":"2022-03-10T17:26:28.105+0100","message":"Local scheduler finished successfully"}
{"level":"INFO","time":"2022-03-10T17:26:28.105+0100","message":"Starting worker pool"}
{"level":"INFO","time":"2022-03-10T17:26:28.105+0100","message":"Starting worker pool with capacity of 30 workers"}
{"level":"INFO","time":"2022-03-10T17:26:29.107+0100","message":"Worker pool assigned worker to reconcile component 'CRDs' on cluster 'local' (OperationEntity [SchedulingID=local--e97696b5-65a9-469e-ba39-fb8fdb11cd3c,CorrelationID=local--e4c027ca-058d-4280-bcf3-0619e65c72e7,RuntimeID=local,ClusterConfig=1,Component=CRDs,Prio=1,State=new,Type=reconcile])"}
{"level":"INFO","time":"2022-03-10T17:26:29.107+0100","message":"Worker pool assigned 1 of 1 processable operations to workers"}
{"level":"INFO","time":"2022-03-10T17:26:29.112+0100","message":"Component 'CRDs' has status 'running'"}
...
{"level":"INFO","time":"2022-03-10T17:26:41.987+0100","message":"Runner: reconciliation of component 'CRDs' for version 'main' finished successfully"}
{"level":"INFO","time":"2022-03-10T17:26:41.987+0100","message":"Component 'CRDs' has status 'success'"}
{"level":"INFO","time":"2022-03-10T17:26:43.106+0100","message":"Worker pool assigned worker to reconcile component 'eventing' on cluster 'local' (OperationEntity [SchedulingID=local--e97696b5-65a9-469e-ba39-fb8fdb11cd3c,CorrelationID=local--4f260131-778c-4ea0-8d76-c9959b472702,RuntimeID=local,ClusterConfig=1,Component=eventing,Prio=2,State=new,Type=reconcile])"}
{"level":"INFO","time":"2022-03-10T17:26:43.106+0100","message":"Worker pool assigned 1 of 1 processable operations to workers"}
{"level":"INFO","time":"2022-03-10T17:26:43.106+0100","message":"Component 'eventing' has status 'running'"}
{"level":"INFO","time":"2022-03-10T17:26:43.108+0100","message":"Cloning GIT repository 'https://github.com/kyma-project/kyma' with revision '1.24.7' into workspace '.workspace/1.24.7'"}
{"level":"INFO","time":"2022-03-10T17:26:43.108+0100","message":"Not able to create an In Cluster Client"}
{"level":"INFO","time":"2022-03-10T17:27:05.612+0100","message":"Removing nats-operator chart resources","action":"pre-action","version":"main","step":"removeNatsOperator"}
{"level":"INFO","time":"2022-03-10T17:27:05.691+0100","message":"Cannot add the resource kind: ServiceAccount, name: nats-operator to watchables","action":"pre-action","version":"main","step":"removeNatsOperator"}
{"level":"INFO","time":"2022-03-10T17:27:05.691+0100","message":"Deleting: kind: ServiceAccount, name: nats-operator, namespace: kyma-system","action":"pre-action","version":"main","step":"removeNatsOperator"}
{"level":"INFO","time":"2022-03-10T17:27:06.050+0100","message":"Cannot add the resource kind: ServiceAccount, name: nats-server to watchables","action":"pre-action","version":"main","step":"removeNatsOperator"}
{"level":"INFO","time":"2022-03-10T17:27:06.050+0100","message":"Deleting: kind: ServiceAccount, name: nats-server, namespace: kyma-system","action":"pre-action","version":"main","step":"removeNatsOperator"}
{"level":"INFO","time":"2022-03-10T17:27:06.127+0100","message":"Cannot add the resource kind: ClusterRole, name: nats-operator to watchables","action":"pre-action","version":"main","step":"removeNatsOperator"}
{"level":"INFO","time":"2022-03-10T17:27:06.127+0100","message":"Deleting: kind: ClusterRole, name: nats-operator, namespace: kyma-system","action":"pre-action","version":"main","step":"removeNatsOperator"}
{"level":"INFO","time":"2022-03-10T17:27:06.188+0100","message":"Cannot add the resource kind: ClusterRole, name: nats-server to watchables","action":"pre-action","version":"main","step":"removeNatsOperator"}
{"level":"INFO","time":"2022-03-10T17:27:06.188+0100","message":"Deleting: kind: ClusterRole, name: nats-server, namespace: kyma-system","action":"pre-action","version":"main","step":"removeNatsOperator"}
{"level":"INFO","time":"2022-03-10T17:27:06.248+0100","message":"Cannot add the resource kind: ClusterRoleBinding, name: nats-operator-binding to watchables","action":"pre-action","version":"main","step":"removeNatsOperator"}
{"level":"INFO","time":"2022-03-10T17:27:06.248+0100","message":"Deleting: kind: ClusterRoleBinding, name: nats-operator-binding, namespace: kyma-system","action":"pre-action","version":"main","step":"removeNatsOperator"}
{"level":"INFO","time":"2022-03-10T17:27:06.310+0100","message":"Cannot add the resource kind: ClusterRoleBinding, name: nats-server-binding to watchables","action":"pre-action","version":"main","step":"removeNatsOperator"}
{"level":"INFO","time":"2022-03-10T17:27:06.310+0100","message":"Deleting: kind: ClusterRoleBinding, name: nats-server-binding, namespace: kyma-system","action":"pre-action","version":"main","step":"removeNatsOperator"}
{"level":"INFO","time":"2022-03-10T17:27:06.420+0100","message":"Deleting: kind: Deployment, name: nats-operator, namespace: kyma-system","action":"pre-action","version":"main","step":"removeNatsOperator"}
{"level":"INFO","time":"2022-03-10T17:27:06.464+0100","message":"Cannot add the resource kind: DestinationRule, name: eventing-nats to watchables","action":"pre-action","version":"main","step":"removeNatsOperator"}
{"level":"INFO","time":"2022-03-10T17:27:06.464+0100","message":"Deleting: kind: DestinationRule, name: eventing-nats, namespace: kyma-system","action":"pre-action","version":"main","step":"removeNatsOperator"}
{"level":"INFO","time":"2022-03-10T17:27:06.583+0100","message":"Cannot add the resource kind: NatsCluster, name: eventing-nats to watchables","action":"pre-action","version":"main","step":"removeNatsOperator"}
{"level":"INFO","time":"2022-03-10T17:27:06.583+0100","message":"Deleting: kind: NatsCluster, name: eventing-nats, namespace: kyma-system","action":"pre-action","version":"main","step":"removeNatsOperator"}
{"level":"INFO","time":"2022-03-10T17:27:06.614+0100","message":"Removing nats-operator CRDs","action":"pre-action","version":"main","step":"removeNatsOperator"}
{"level":"INFO","time":"2022-03-10T17:27:06.683+0100","message":"Removing orphaned NATS pods","action":"pre-action","version":"main","step":"removeNatsOperator"}
{"level":"INFO","time":"2022-03-10T17:27:06.790+0100","message":"Step skipped","action":"pre-action","version":"main","step":"MigrateEventTypePrefixConfig","reason":"Eventing controller deployment is already managed by Kyma reconciler"}
{"level":"INFO","time":"2022-03-10T17:27:13.107+0100","message":"Component 'eventing' has status 'running'"}
{"level":"INFO","time":"2022-03-10T17:27:43.109+0100","message":"Component 'eventing' has status 'running'"}
{"level":"INFO","time":"2022-03-10T17:28:13.110+0100","message":"Component 'eventing' has status 'running'"}
{"level":"INFO","time":"2022-03-10T17:28:43.110+0100","message":"Component 'eventing' has status 'running'"}
{"level":"INFO","time":"2022-03-10T17:29:13.111+0100","message":"Component 'eventing' has status 'running'"}
{"level":"INFO","time":"2022-03-10T17:29:43.101+0100","message":"Component 'eventing' has status 'running'"}
{"level":"INFO","time":"2022-03-10T17:30:13.083+0100","message":"Component 'eventing' has status 'running'"}
{"level":"INFO","time":"2022-03-10T17:30:43.081+0100","message":"Component 'eventing' has status 'running'"}
{"level":"INFO","time":"2022-03-10T17:31:12.547+0100","message":"Runner: reconciliation of component 'eventing' for version 'main' finished successfully"}
{"level":"INFO","time":"2022-03-10T17:31:12.547+0100","message":"Component 'eventing' has status 'success'"}
{"level":"INFO","time":"2022-03-10T17:31:14.078+0100","message":"Stopping worker pool"}
{"level":"INFO","time":"2022-03-10T17:31:14.079+0100","message":"Worker pool finished successfully"}

took 5m 12s
```

## Validate jetstream is enabled now

```shell
# check that jetstream section is inside the config
$ k get cm -n kyma-system eventing-nats-config -oyaml | grep -C 5 jetstream
    ###################################
    #                                 #
    # NATS JetStream                  #
    #                                 #
    ###################################
    jetstream {
      max_mem: 1Gi
    }
    debug: true
    trace:  true
    lame_duck_duration: 120s

# ENABLE_JETSTREAM_BACKEND env var is set on the eventing-controller
$ k get deployments.apps -n kyma-system eventing-controller|grep -i -A 1 ENABLE_JETSTREAM_BACKEND
empty

# Check sts events (we need this to make sure a new pod is created when enabling jetstream)
$ k describe sts -n kyma-system  eventing-nats|grep -A 10 Events
Events:
  Type    Reason            Age                   From                    Message
  ----    ------            ----                  ----                    -------
  Normal  SuccessfulDelete  8m1s                  statefulset-controller  delete Pod eventing-nats-0 in StatefulSet eventing-nats successful
  Normal  SuccessfulCreate  5m59s (x2 over 146m)  statefulset-controller  create Pod eventing-nats-0 in StatefulSet eventing-nats successful


# check start time of nats pod (to make sure it got recreated later)
$ k get pods -n kyma-system eventing-nats-0 -o jsonpath='{.status.startTime}'
2022-03-10T16:29:13Z%

# nats CLI can show stream info
$ kubectl -n kyma-system port-forward svc/eventing-nats 4222
$ nats stream ls
 nats stream ls
╭───────────────────────────────────────────────────────────────────────────╮
│                                  Streams                                  │
├──────┬─────────────┬─────────────────────┬──────────┬──────┬──────────────┤
│ Name │ Description │ Created             │ Messages │ Size │ Last Message │
├──────┼─────────────┼─────────────────────┼──────────┼──────┼──────────────┤
│ sap  │             │ 2022-03-10 16:03:01 │ 0        │ 0 B  │ never        │
╰──────┴─────────────┴─────────────────────┴──────────┴──────┴──────────────╯

# publish event
$ kyma-send-ce-structured ns213321.kymatunas.shoot.canary.k8s-hana.ondemand.com
< HTTP/2 204
...

$ nats stream ls
╭───────────────────────────────────────────────────────────────────────────╮
│                                  Streams                                  │
├──────┬─────────────┬─────────────────────┬──────────┬──────┬──────────────┤
│ Name │ Description │ Created             │ Messages │ Size │ Last Message │
├──────┼─────────────┼─────────────────────┼──────────┼──────┼──────────────┤
│ sap  │             │ 2022-03-10 16:03:01 │ 0        │ 0 B  │ 23.65s       │
╰──────┴─────────────┴─────────────────────┴──────────┴──────┴──────────────╯



```

**Comment by nachtmaar** [comment]
New update

- Upgrade from main (jetstream=disabled) to main (jetstream=enabled):
	 1. kyma deploy with **stable** (stable-9407a4a) version `works`
	 2. Kyma deploy with **unstable** version (main-d5613a0) does `not work`
	 3. Using mothership-reconciler (commit 654fde13d9284e5c9e8c05ea9b9297b87061d034) does `not work`
	 4. Using mothership-reconciler (commit 654fde13d9284e5c9e8c05ea9b9297b87061d034) and **nats config reloader** `works`

**Comment by nachtmaar** [comment]
For anyone reading this issue, this is how I did the upgrade/downgrade/testing.

# How to 

## Perform the upgrade

`upgrade.sh`:

```bash
#!/usr/bin/env bash
set -x

# kyma-rec-mothership local --components eventing --version main --value global.features.enableJetStream=true
rm -r .workspace/kyma-13321-handle-jetstream-flag-reconciler || true
set -e
kyma-rec-mothership local --components eventing --version kyma-13321-handle-jetstream-flag-reconciler --value global.features.enableJetStream=true --git https://github.com/nachtmaar/kyma -v
set +e
./check-jetstream.sh
```

`downgrade.sh`:
```bash
#!/usr/bin/env bash
set -x

kubectl scale --replicas 0 -n kyma-system deployment eventing-controller
kubectl delete sts -n kyma-system eventing-nats
kubectl set env -n kyma-system deployment/eventing-controller  ENABLE_JETSTREAM_BACKEND=false
kubectl scale --replicas 1 -n kyma-system deployment eventing-controller

rm -r .workspace/main || true
kyma-rec-mothership local --components eventing --version main --value global.features.enableJetStream=false
./check-jetstream.sh

```

## Test that the upgrade worked

`check-jetstream.sh`:

```bash
#!/usr/bin/env bash
echo "ectrl: deployment"
kubectl get deployments.apps -n kyma-system eventing-controller -oyaml |grep -i -A 1 ENABLE_JETSTREAM_BACKEND
echo "logs: eventing-nats"
kubectl logs -n kyma-system eventing-nats-0|grep Starting || kubectl logs -n kyma-system eventing-nats-0 -c nats |grep Starting
echo "nats pod containers:"
kubectl get pod -n kyma-system eventing-nats-0 -ojsonpath='{.spec.containers}' | jq -r ".[] | .name"
echo "cm: eventing-nats-config":
kubectl get cm -n kyma-system eventing-nats-config -ojsonpath='{.data.nats\.conf}' | head -n 30

```

**Comment by nachtmaar** [comment]
Upgrade takes `4:02.78 total` (using upgrade.sh)
